File: docs/JAMIA/comparisons/RPython/TM/31745509_str_contains_nullable.ts
/**
 * RPython SO#31745509 — Using str.contains on pandas dataframe with nullable column
 * Effect: Crash
 * Bug class: Nullable type
 *
 * In pandas, str.contains on a column with NaN returns NaN for those rows.
 * Applying ~ (NOT) on NaN crashes. The user must handle nulls explicitly.
 *
 * In tidy-ts, a nullable string column is typed as string | null. Calling
 * string methods without null-checking is a compile-time error.
 */
import { createDataFrame } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { v: "File corruption" as string | null },
  { v: "Registry error" },
  { v: null },
  { v: "File missing" },
  { v: "Other issue" },
]);

// @ts-expect-error — 'v' is possibly 'null'
const wrong = df.filter((r) => !r.v.includes("File"));

// Fix: null-check first
const filtered = df.filter((r) => r.v !== null && !r.v.includes("File") && !r.v.includes("Registry"));
filtered.print();

File: docs/JAMIA/comparisons/RPython/TM/31745509_str_contains_nullable.py
# ID: SO#31745509
# Language: Python
# Bug class: Nullable
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: str.contains on nullable column returns NaN, bitwise NOT fails on NaN. Missing values break operations.
# Reproduction status: Fixed (pandas 2.x handles nullable StringDtype)
# Type system catch: `number` rejects `number | null`

import pandas as pd
import numpy as np

df = pd.DataFrame({
    'V': ['File corruption', 'Registry error', np.nan, 'File missing', 'Other issue'],
})

# BUG: str.contains returns NaN for null rows, ~ fails on NaN
try:
    result = df[~df['V'].str.contains("File|Registry", na=False)]
    # With na=False it works — but without it:
    mask = df['V'].str.contains("File|Registry")
    print(f"mask with NaN: {mask.tolist()}")
    filtered = df[~mask]  # This crashes due to NaN in mask
    print(filtered)
except (TypeError, ValueError) as e:
    print(f"Crash: {e}")
